### DISCUSSION TITLE
XCode Cloud

### DISCUSSION THREAD

**Discussion by aschmitt** [MobileNativeFoundation/discussions#151] [General]
Hi there,

has anyone looked into XCode Cloud and can compare it to other cloud providers such as Bitrise?

Thanks

**Comment by keith** [comment]
I've only watched the WWDC sessions so far, but I think there are some clear limitations that would block us from using it in general, like versioning the CI configuration in our repo, and dynamically deciding what jobs to spin up given the specifics of a change (we could probably implement something like this with their API, but with buildkite this works really well out of the box)

**Reply by jschear** [reply]
> versioning the CI configuration in our repo

Is that not supported in Xcode Cloud, or is that the only option?

**Reply by keith** [reply]
AFAICT the workflows are entirely configured as server side config https://developer.apple.com/documentation/Xcode/Configuring-Your-First-Xcode-Cloud-Workflow I think you can even edit them in iTC

**Reply by jschear** [reply]
Ouch, yeah -- definitely gets tricky to coordinate changes if the configuration lives outside of the repo.

**Reply by skofgar** [reply]
While the job definitions themselves seem to be server-side, I am able to create [pre and post scripts](https://developer.apple.com/documentation/xcode/writing-custom-build-scripts). I know that doesn't allow versioning the entire workflow, but at least some can be tracked in source.

I use the scripts to install CocoaPods dependencies and download some files.

**Comment by skofgar** [comment]
A few observations of mine:

- One thing I like is that Xcode Cloud build stats, logs, test results and more can be accessed directly from AppStore Connect. 
- When accessing logs I noticed that AppStore Connect and Xcode load the logs piece by piece, so I sometimes have to wait for a minute until I have the entire log present.
- Last year submitting builds to TestFlight/AppStore through Xcode Cloud failed for me, but nowadays it actually works quite well.
